File: react-native/test/test_mode.test.ts
import { GoldenKrillAds } from '../src/ads';
import { GoldenKrillClient } from '../src/client';
import { adsUrl, configUrl } from '../src/endpoints';
import { enc, resp, memStore, flush } from './_helpers';

// Prompt 16 test mode (RN parity): testMode defaults to __DEV__, sends test=1 on /ads +
// /config and test:true on beacons, and renders the server's TEST AD normally.
function recordingAds(testMode?: boolean) {
  const urls: string[] = [];
  let body: any = null;
  const ads = new GoldenKrillAds('com.x', {
    storage: memStore(),
    now: () => 1e12,
    testMode,
    fetchImpl: async (u: any, init: any) => {
      urls.push(String(u));
      if (init && init.body) { body = JSON.parse(init.body); return resp(200, ''); }
      if (String(u).includes('/config/')) return resp(200, enc({ house_cooldown_sec: 0, reserve_share: false }));
      return resp(200, enc({ a: [[0, 'i', 's']], o: [], n: 'N', t: 1 }));
    },
  });
  return { ads, urls: () => urls, body: () => body };
}

async function run(r: ReturnType<typeof recordingAds>) {
  await r.ads.ensureReady('banner');
  await r.ads.fallbackAd('banner'); // records an impression -> beacon
  await flush();
}

describe('test mode (RN)', () => {
  it('endpoints add test=1 only when requested', () => {
    expect(adsUrl('p', 'banner', 'en', undefined, undefined, true)).toContain('test=1');
    expect(adsUrl('p', 'banner')).not.toContain('test=1');
    expect(configUrl('p', undefined, true)).toContain('test=1');
    expect(configUrl('p')).not.toContain('test=1');
  });

  it('testMode sends test=1 on ads + config and test:true on the beacon', async () => {
    const r = recordingAds(true);
    await run(r);
    expect(r.urls().some((u) => u.includes('/config/') && u.includes('test=1'))).toBe(true);
    expect(r.urls().some((u) => u.includes('/ads?') && u.includes('test=1'))).toBe(true);
    expect(r.body().test).toBe(true);
  });

  it('explicit testMode=false never sends test', async () => {
    const r = recordingAds(false);
    await run(r);
    expect(r.urls().every((u) => !u.includes('test=1'))).toBe(true);
    expect(r.body() && r.body().test).toBeFalsy();
  });

  it('client default is false; ads layer defaults to __DEV__', () => {
    expect(new GoldenKrillClient('p', {}).testMode).toBe(false);
    (global as any).__DEV__ = true;
    expect((recordingAds(undefined).ads as any).client.testMode).toBe(true);
    (global as any).__DEV__ = false;
    expect((recordingAds(undefined).ads as any).client.testMode).toBe(false);
  });
});

File: flutter/test/test_mode_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:goldenkrill/goldenkrill.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:shared_preferences/shared_preferences.dart';

// Prompt 16 test mode (Flutter): testMode defaults to kDebugMode, sends test=1 on /ads +
// /config and test:true on beacons, and renders the server's TEST AD normally.
void main() {
  TestWidgetsFlutterBinding.ensureInitialized();
  setUp(() => SharedPreferences.setMockInitialValues({}));

  String gk1(Object json) => CatalogCodec.encode(jsonEncode(json));

  MockClient mock(List<Uri> urls, void Function(Map<String, dynamic>)? onEvents) =>
      MockClient((req) async {
        urls.add(req.url);
        if (req.url.path.contains('/config/')) {
          return http.Response(gk1({'house_cooldown_sec': 0, 'reserve_share': false}), 200);
        }
        if (req.url.path.contains('/events')) {
          onEvents?.call(jsonDecode(req.body) as Map<String, dynamic>);
          return http.Response('', 200);
        }
        return http.Response(gk1({'a': [[0, 'i', 's']], 'o': <dynamic>[], 'n': 'N', 't': 1}), 200);
      });

  Future<void> settle() => Future<void>.delayed(const Duration(milliseconds: 50));

  test('testMode sends test=1 on ads + config and test:true on the beacon', () async {
    final urls = <Uri>[];
    Map<String, dynamic>? body;
    final ads = GoldenKrillAds(
        client: GoldenKrillClient(package: 'com.x', client: mock(urls, (b) => body = b), testMode: true));
    await ads.ensureReady(slot: 'banner');
    await ads.fallbackAd('banner');
    await settle();
    expect(urls.any((u) => u.path.contains('/config/') && u.query.contains('test=1')), isTrue);
    expect(urls.any((u) => u.path.endsWith('/ads') && u.queryParameters['test'] == '1'), isTrue);
    expect(body?['test'], true);
  });

  test('no test=1 and no test flag when testMode is false', () async {
    final urls = <Uri>[];
    Map<String, dynamic>? body;
    final ads = GoldenKrillAds(
        client: GoldenKrillClient(package: 'com.x', client: mock(urls, (b) => body = b), testMode: false));
    await ads.ensureReady(slot: 'banner');
    await ads.fallbackAd('banner');
    await settle();
    expect(urls.every((u) => !u.query.contains('test=1')), isTrue);
    expect(body?.containsKey('test'), isFalse);
  });

  test('GoldenKrillAds defaults testMode to kDebugMode', () {
    // Exercises the default client path (testMode ?? kDebugMode) without a network call.
    expect(GoldenKrillAds(package: 'com.x'), isNotNull);
    expect(GoldenKrillAds(package: 'com.x', testMode: false), isNotNull);
  });
}
